### DISCUSSION TITLE
[Feature request] How to setup reverse DNS and add an A record

### DISCUSSION THREAD

**Discussion by ghost** [AdguardTeam/AdGuardHome#2731] [Q&A]
### Prerequisites

- [x] I am running the latest version
- [x] I checked the documentation and found no answer
- [x] I checked to make sure that this issue has not already been filed

### Problem Description

How to setup reverse DNS and add an A record

**Comment by ameshkov** [comment]
Could you please be more specific?

Reverse DNS and A records are a little bit different things

**Comment by ghost** [comment]
sorry,it's two functions
1. setup reverse DNS
2. add an A record

**Comment by ghost** [comment]
eg

[Configuring Reverse DNS in BIND 9](https://philchen.com/2007/04/04/configuring-reverse-dns)

[add A record in bind9](https://serverfault.com/questions/903827/how-to-add-a-record-in-bind9)

**Comment by ameshkov** [comment]
> setup reverse DNS

Not possible in the current version, but that's a good case for #2102.

@ainar-g would it be possible with $dnsrewrite in it's current implementation?

> add an A record

You can use "Filters -> DNS rewrites" settings for that.

**Reply by jeffr1975** [reply]
DNS rewrites cause major problems with nginx reverse proxy with subdomains.  A records on pihole work fine.  I can't find a nginx config that allows it to handle dns rewrites.  Just having support for A records would be very good.

**Reply by ameshkov** [reply]
I honestly have no idea how DNS rewrites and nginx are connected.

Basically, when you add a DNS rewrite that looks like `example.org -> 127.0.0.1`, you are adding an A record.

If you want a more fine-grained control over it, you can use $dnsrewrite rules:
https://github.com/AdguardTeam/AdGuardHome/wiki/Hosts-Blocklists#dnsrewrite

**Comment by ainar-g** [comment]
@ameshkov, you mean, the Reverse DNS part?  Not currently, but overall the extension seems to be trivial.  Maybe for v0.106.0?  We have enough on our plate for v0.105.0 currently.

**Comment by ameshkov** [comment]
@ainar-g why can't I add `192.168.0.1.in-addr.arpa$dnsrewrite=NOERROR:PTR:hostname`?

**Comment by ainar-g** [comment]
@ameshkov, because different record types have different types of values, and we decode them into [RRValue]s correspondingly.  And we also need to catch the `CNAME` exception at the same stage.

https://github.com/AdguardTeam/urlfilter/blob/d0665c13ffaaba988fd29def346ac032c7eeb4bd/rules/dnsrewrite.go#L129-L168

We could just put the value as verbatim text in the `default` case, but that means that when we **do** decode the value properly in the future, we may break some clients.

[RRValue]: https://pkg.go.dev/github.com/AdguardTeam/urlfilter/rules#RRValue

**Comment by ameshkov** [comment]
@ainar-g it should be explained in the documentation that some RRTypes don't support values **yet**.

There're some that are rather important and would be useful to have in the first release:

* `HTTPS` - this record type may be really useful. For instance, people may use it to achieve "HTTPS everywhere"-like behavior.
* `PTR` - for configuring hostnames on the network. May also be useful to setup bonjour-like services.
* `MX` - for configuring mail servers on the network.

Please check how complicated is it to add those.

**Comment by ainar-g** [comment]
@ameshkov, `PTR` looks easy enough.  Do we need an ability to set `MX`'s [`PREFERENCE` parameter]?  If no, also easy; if yes, medium-easy, once we agree on a format.

As far as I can see, `HTTPS` is basically the same as `SVCB`, so we could technically get both in one go.  Again, we need to agree on the syntax first.  After that, the difficulty seems to be medium.

[`PREFERENCE` parameter]: https://tools.ietf.org/html/rfc1035#section-3.3.9

**Comment by ameshkov** [comment]
Here's how I suggest to deal with MX:

`|google.com^$dnsrewrite=NOERROR:MX:30 alt2.aspmx.l.google.com.`

```
;; ANSWER SECTION:
google.com.		187	IN	MX	30 alt2.aspmx.l.google.com.
```

No complicated stuff is needed.

**Comment by ainar-g** [comment]
Should be possible as of v0.105.0-beta.2.  @da-gongren, can you please update and check?

[Wiki docs](https://github.com/AdguardTeam/AdGuardHome/wiki/Hosts-Blocklists#dnsrewrite).
